File: sbt-plugin/src/main/scala/com/monadiccloud/bindingz/contract/plugin/sbt/resources/SourceCodeFactory.scala
/*
 * Copyright (c) 2019 Connor Goulding
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.monadiccloud.bindingz.contract.plugin.sbt.resources

import argus.macros.ModelBuilder
import argus.schema.Schema

object SourceCodeFactory {

  private val modelBuilder = new ModelBuilder[scala.reflect.runtime.universe.type](scala.reflect.runtime.universe)

  def create(content: String, contractName: String, packageName: Option[String]) = {
    val schema = Schema.fromJson(content)

    val (_, trees) = modelBuilder.mkSchemaDef(contractName, schema)

    val packageCode = packageName.map(pn => s"package $pn;\n\n").getOrElse("")

    // Clean up the code a little to make it more readable
    packageCode + trees.map(tree => scala.reflect.runtime.universe.showCode(tree)).mkString("\n\n")
  }
}

File: source-code-provider/src/main/java/com/monadiccloud/bindingz/source/code/SourceCodeFactory.java
/*
 * Copyright (c) 2019 Connor Goulding
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.monadiccloud.bindingz.source.code;

import com.monadiccloud.bindingz.contract.core.configuration.SourceCodeConfiguration;
import com.monadiccloud.bindingz.contract.core.model.ContractDto;
import com.monadiccloud.bindingz.contract.core.model.SourceDto;
import com.monadiccloud.bindingz.source.code.provider.jsonschematopojo.JsonSchema2PojoProvider;

import java.util.List;

public class SourceCodeFactory {

    public List<SourceDto> create(ContractDto contractDto, SourceCodeConfiguration configuration) throws SourceCodeProviderException {
        switch (SourceCodeProviderType.valueOf(configuration.getSourceCodeProvider())) {
            case JSONSCHEMA2POJO: {
                return new JsonSchema2PojoProvider().create(contractDto, configuration);
            }
            default:
                throw new UnsupportedOperationException("Unknown provider type: " + configuration.getSourceCodeProvider());
        }
    }
}
